README.md
# youtube-history-analysis

[![PyPI](https://img.shields.io/pypi/v/youtube-history-analysis?style=flat-square)](https://pypi.python.org/pypi/youtube-history-analysis/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/youtube-history-analysis?style=flat-square)](https://pypi.python.org/pypi/youtube-history-analysis/)
[![PyPI - License](https://img.shields.io/pypi/l/youtube-history-analysis?style=flat-square)](https://pypi.python.org/pypi/youtube-history-analysis/)
[![Coookiecutter - Wolt](https://img.shields.io/badge/cookiecutter-Wolt-00c2e8?style=flat-square&logo=cookiecutter&logoColor=D4AA00&link=https://github.com/woltapp/wolt-python-package-cookiecutter)](https://github.com/woltapp/wolt-python-package-cookiecutter)


---

**Documentation**: [https://armanckeser.github.io/youtube-history-analysis](https://armanckeser.github.io/youtube-history-analysis)

**Source Code**: [https://github.com/armanckeser/youtube-history-analysis](https://github.com/armanckeser/youtube-history-analysis)

**PyPI**: [https://pypi.org/project/youtube-history-analysis/](https://pypi.org/project/youtube-history-analysis/)

---

See how your YouTube interests evolved over time
![Example](/docs/percentage_per_month_dark.png)

## Installation

```sh
python -m venv yt-history-venv
# On Windows ./yt-history-venv/Scripts/activate.bat
# On MAC source ./yt-history-venv/bin/activate
pip install youtube-history-analysis
```

## Usage
### Get a YouTube API Key

1. Visit the [Google Cloud Console](https://console.cloud.google.com/).
2. Click the project drop-down and select or create the project for which you want to add an API key.
3. Click the hamburger menu and select APIs & Services > Credentials.
4. On the Credentials page, click Create credentials > API key.
5. The API key created dialog displays your newly created API key.

Remember to restrict the API key so that it can only be used with certain websites or IP addresses by clicking the Edit button for the API key and then setting the restrictions in the Key restriction section.
### Get your YouTube History as JSON
1. Visit [Google Takeout](https://takeout.google.com/) and sign in to your Google account.
2. Scroll down to the "YouTube" section and click All data included.
3. Click the Deselect All button and then select the checkbox next to Watch history.
4. Click the Next button at the bottom of the page.
5. On the next page, you can select the file type and delivery method for your takeout. Make sure to select JSON as the file type.
6. Click the Create export button to start the export process.

Once the export is complete, you will receive an email with a link to download your takeout. The downloaded file will be a zip archive containing your YouTube watch history in JSON format.

```sh
python -m youtube_history_analysis $API_KEY --watch-history-file-path $WATCH_HISTORY_JSON_PATH
```

This will create an `outputs` folder with a bunch of `.csv` files and a few `.png` files. Feel free to use the `.csv` file to do your own analysis.

## Development

* Clone this repository
* Requirements:
  * [Poetry](https://python-poetry.org/)
  * Python 3.9+
* Create a virtual environment and install the dependencies

```sh
poetry install
```

* Activate the virtual environment

```sh
poetry shell
```

### Testing

```sh
pytest
```

### Documentation

The documentation is automatically generated from the content of the [docs directory](./docs) and from the docstrings
 of the public signatures of the source code. The documentation is updated and published as a [Github project page
 ](https://pages.github.com/) automatically as part each release.

### Releasing

Trigger the [Draft release workflow](https://github.com/armanckeser/youtube-history-analysis/actions/workflows/draft_release.yml)
(press _Run workflow_). This will update the changelog & version and create a GitHub release which is in _Draft_ state.

Find the draft release from the
[GitHub releases](https://github.com/armanckeser/youtube-history-analysis/releases) and publish it. When
 a release is published, it'll trigger [release](https://github.com/armanckeser/youtube-history-analysis/blob/master/.github/workflows/release.yml) workflow which creates PyPI
 release and deploys updated documentation.

### Pre-commit

Pre-commit hooks run all the auto-formatters (e.g. `black`, `isort`), linters (e.g. `mypy`, `flake8`), and other quality
 checks to make sure the changeset is in good shape before a commit/push happens.

You can install the hooks with (runs for each commit):

```sh
pre-commit install
```

Or if you want them to run only for each push:

```sh
pre-commit install -t pre-push
```

Or if you want e.g. want to run all checks manually for all files:

```sh
pre-commit run --all-files
```

---

This project was generated using the [wolt-python-package-cookiecutter](https://github.com/woltapp/wolt-python-package-cookiecutter) template.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.2] - 2022-12-09
### Added
- Added prints of script state for better user experience

## [1.0.1] - 2022-12-08
### Changed
- Changed README.md and added more documentation

## [1.0.0] - 2022-12-08
### Added
- Added main function that creates a few analysis files
- Use `poetry run python -m youtube_history_analysis  $API_KEY $WATCH_HISTORY_FILE_PATH` to create analysis

[Unreleased]: https://github.com/armanckeser/youtube-history-analysis/compare/1.0.2...master
[1.0.2]: https://github.com/armanckeser/youtube-history-analysis/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/armanckeser/youtube-history-analysis/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/armanckeser/youtube-history-analysis/tree/1.0.0


src/youtube_history_analysis/main.py
import json
import os
from typing import TYPE_CHECKING

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import typer
from googleapiclient.discovery import build

if TYPE_CHECKING:
    from googleapiclient._apis.youtube.v3.resources import YouTubeResource


YT_URL = "https://www.googleapis.com/youtube/v3/videos"
WATCH_HISTORY_FILE_PATH = "./watch-history.json"


def main(
    youtube_api_key: str, watch_history_file_path: str = WATCH_HISTORY_FILE_PATH
) -> None:
    """Given a YT v3 API key and a watch history JSON filepath, creates analysis graphs under outputs folder

    Args:
        youtube_api_key (str): YouTube api key from https://console.cloud.google.com/apis/dashboard
        watch_history_file_path (str, optional): Path to watch history downloaded from Google Takeout as JSON. Defaults to ./watch-history.json.
    """
    service: YouTubeResource = build("youtube", "v3", developerKey=youtube_api_key)
    videos_collection = service.videos()
    video_categories_collection = service.videoCategories()

    if not os.path.exists("./outputs"):
        os.mkdir("./outputs")

    with open(WATCH_HISTORY_FILE_PATH, encoding="utf-8") as watch_history_file:
        watch_history = json.load(watch_history_file)

    watched_urls = {}
    for watched in watch_history:
        url = watched.get("titleUrl", None)
        title = watched.get("title", None)[8:]
        if url:
            watched_urls[title] = url

    unique_watched_urls = list(set(watched_urls.values()))
    unique_watched_ids = [url[32:] for url in unique_watched_urls]

    watched_video_snippets = []
    for i in range(0, len(unique_watched_ids), 50):
        video_collection_request = videos_collection.list(
            part="snippet", id=unique_watched_ids[i : i + 50]  # noqa: E203
        )
        snippets_response = video_collection_request.execute()
        snippets = [item["snippet"] for item in snippets_response["items"]]
        watched_video_snippets += snippets

    unique_category_ids = list(
        {snippet["categoryId"] for snippet in watched_video_snippets}
    )
    video_categories_request = video_categories_collection.list(
        part="snippet", id=unique_category_ids
    )
    video_categories_response = video_categories_request.execute()
    video_category_id_to_title = {}
    for item in video_categories_response["items"]:
        video_category_id_to_title[item["id"]] = item["snippet"]["title"]

    df_snippets = pd.DataFrame(watched_video_snippets)
    df_watch_history = pd.DataFrame(watch_history)
    df_watch_history.to_csv("outputs/watch_history.csv")
    df_snippets.to_csv("outputs/snippets.csv")
    df_watch_history.title = df_watch_history.title.apply(lambda title: title[8:])  # type: ignore
    df_merged = df_watch_history.merge(df_snippets, on="title", how="left")
    df_merged["categoryTitle"] = df_merged["categoryId"].apply(
        lambda id: video_category_id_to_title.get(id, "NaN")
    )
    df_merged.drop(
        [
            "subtitles",
            "products",
            "activityControls",
            "header",
            "thumbnails",
            "liveBroadcastContent",
        ],
        axis=1,
    )
    df_merged.to_csv("outputs/full_history.csv")
    df_merged["time"] = pd.to_datetime(df_merged["time"])  # type: ignore
    df_merged["dayWatched"] = pd.to_datetime(  # type: ignore
        df_merged["time"].apply(lambda time: str(time)[:10])
    )
    df_merged.set_index("title")
    df_category_to_day = df_merged.filter(items=["dayWatched", "categoryTitle"])
    ctdf = (
        df_category_to_day.reset_index()
        .groupby(["dayWatched", "categoryTitle"], as_index=False)
        .count()
        # rename isn't strictly necessary here, it's just for readability
        .rename(columns={"index": "count"})
    )
    pivotted = ctdf.pivot_table("count", "dayWatched", "categoryTitle")  # type: ignore
    df_pivotted_by_month = pivotted.resample("M").sum()
    percentages = df_pivotted_by_month.div(df_pivotted_by_month.sum(axis=1), axis=0)

    plt.style.use("dark_background")  # type: ignore
    vals = np.linspace(0, 1, 22)
    np.random.shuffle(vals)
    cmap = plt.cm.colors.ListedColormap(plt.cm.gnuplot(vals))  # type: ignore
    ax = percentages.plot.area(fontsize=12, figsize=(20, 10), cmap=cmap)
    ax.set_xlabel("Date")
    ax.set_ylabel("Percentage of Video Watched")
    ax.set_title("My YouTube Interest by Month")
    handles, labels = ax.get_legend_handles_labels()
    ax.legend(
        handles[::-1],
        labels[::-1],
        fontsize=11,
        loc="upper left",
        bbox_to_anchor=(1.0, 1.0),
    )
    ax.get_figure().savefig(os.path.join("outputs", "percentages_per_month.png"))

    ax = df_pivotted_by_month.plot.area(fontsize=13, figsize=(20, 10))
    ax.set_xlabel("Date")
    ax.set_ylabel("Number of videos watched")
    ax.set_title("My YouTube Interest by Month")
    handles, labels = ax.get_legend_handles_labels()
    ax.legend(handles[::-1], labels[::-1], fontsize=11, loc="upper left")
    ax.get_figure().savefig(os.path.join("outputs", "numbers_per_month.png"))
    service.close()


if __name__ == "__main__":
    typer.run(main)


src/youtube_history_analysis/__main__.py
import json
import os
from typing import TYPE_CHECKING

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import typer
from googleapiclient.discovery import build

if TYPE_CHECKING:
    from googleapiclient._apis.youtube.v3.resources import YouTubeResource


YT_URL = "https://www.googleapis.com/youtube/v3/videos"
WATCH_HISTORY_FILE_PATH = "./watch-history.json"


def main(
    youtube_api_key: str, watch_history_file_path: str = WATCH_HISTORY_FILE_PATH
) -> None:
    """Given a YT v3 API key and a watch history JSON filepath, creates analysis graphs under outputs folder

    Args:
        youtube_api_key (str): YouTube api key from https://console.cloud.google.com/apis/dashboard
        watch_history_file_path (str, optional): Path to watch history downloaded from Google Takeout as JSON. Defaults to ./watch-history.json.
    """
    print("Analysis Started!")
    print("Connecting to YouTube.")
    service: YouTubeResource = build("youtube", "v3", developerKey=youtube_api_key)
    videos_collection = service.videos()
    video_categories_collection = service.videoCategories()

    if not os.path.exists("./outputs"):
        os.mkdir("./outputs")

    print("Loading watch history.")
    with open(WATCH_HISTORY_FILE_PATH, encoding="utf-8") as watch_history_file:
        watch_history = json.load(watch_history_file)

    watched_urls = {}
    for watched in watch_history:
        url = watched.get("titleUrl", None)
        title = watched.get("title", None)[8:]
        if url:
            watched_urls[title] = url

    unique_watched_urls = list(set(watched_urls.values()))
    unique_watched_ids = [url[32:] for url in unique_watched_urls]

    print("Getting video information from YouTube.")
    watched_video_snippets = []
    for i in range(0, len(unique_watched_ids), 50):
        video_collection_request = videos_collection.list(
            part="snippet", id=unique_watched_ids[i : i + 50]  # noqa: E203
        )
        snippets_response = video_collection_request.execute()
        snippets = [item["snippet"] for item in snippets_response["items"]]
        watched_video_snippets += snippets

    unique_category_ids = list(
        {snippet["categoryId"] for snippet in watched_video_snippets}
    )
    video_categories_request = video_categories_collection.list(
        part="snippet", id=unique_category_ids
    )
    video_categories_response = video_categories_request.execute()
    video_category_id_to_title = {}
    for item in video_categories_response["items"]:
        video_category_id_to_title[item["id"]] = item["snippet"]["title"]

    print("Saving video snippets and watch history information to outputs folder.")
    df_snippets = pd.DataFrame(watched_video_snippets)
    df_watch_history = pd.DataFrame(watch_history)
    df_watch_history.to_csv("outputs/watch_history.csv")
    df_snippets.to_csv("outputs/snippets.csv")
    df_watch_history.title = df_watch_history.title.apply(lambda title: title[8:])  # type: ignore
    df_merged = df_watch_history.merge(df_snippets, on="title", how="left")
    df_merged["categoryTitle"] = df_merged["categoryId"].apply(
        lambda id: video_category_id_to_title.get(id, "NaN")
    )
    df_merged.drop(
        [
            "subtitles",
            "products",
            "activityControls",
            "header",
            "thumbnails",
            "liveBroadcastContent",
        ],
        axis=1,
    )
    df_merged.to_csv("outputs/full_history.csv")
    df_merged["time"] = pd.to_datetime(df_merged["time"])  # type: ignore
    df_merged["dayWatched"] = pd.to_datetime(  # type: ignore
        df_merged["time"].apply(lambda time: str(time)[:10])
    )
    df_merged.set_index("title")
    df_category_to_day = df_merged.filter(items=["dayWatched", "categoryTitle"])
    ctdf = (
        df_category_to_day.reset_index()
        .groupby(["dayWatched", "categoryTitle"], as_index=False)
        .count()
        # rename isn't strictly necessary here, it's just for readability
        .rename(columns={"index": "count"})
    )
    pivotted = ctdf.pivot_table("count", "dayWatched", "categoryTitle")  # type: ignore
    df_pivotted_by_month = pivotted.resample("M").sum()
    percentages = df_pivotted_by_month.div(df_pivotted_by_month.sum(axis=1), axis=0)

    print("Drawing a few example graphs to output folder.")
    plt.style.use("dark_background")  # type: ignore
    vals = np.linspace(0, 1, 22)
    np.random.shuffle(vals)
    cmap = plt.cm.colors.ListedColormap(plt.cm.gnuplot(vals))  # type: ignore
    ax = percentages.plot.area(fontsize=12, figsize=(20, 10), cmap=cmap)
    ax.set_xlabel("Date")
    ax.set_ylabel("Percentage of Video Watched")
    ax.set_title("My YouTube Interest by Month")
    handles, labels = ax.get_legend_handles_labels()
    ax.legend(
        handles[::-1],
        labels[::-1],
        fontsize=11,
        loc="upper left",
        bbox_to_anchor=(1.0, 1.0),
    )
    ax.get_figure().savefig(os.path.join("outputs", "percentages_per_month.png"))

    ax = df_pivotted_by_month.plot.area(fontsize=13, figsize=(20, 10))
    ax.set_xlabel("Date")
    ax.set_ylabel("Number of videos watched")
    ax.set_title("My YouTube Interest by Month")
    handles, labels = ax.get_legend_handles_labels()
    ax.legend(handles[::-1], labels[::-1], fontsize=11, loc="upper left")
    ax.get_figure().savefig(os.path.join("outputs", "numbers_per_month.png"))
    service.close()

    print(
        "Done! Go check out your analysis in the outputs folder!"
        "You can also run some on your own using the csv files."
        "Thanks for using youtube-history-analysis"
    )


if __name__ == "__main__":
    typer.run(main)
